Fix TypeScript and run it on CI

Gcommer opened the pull request:
* Remove ".index" from a few route -- they still seem to work the same, but this fixes some path strings that TypeScript didn't like because they needed a weird trailing slash.
* Update video-react types to have this fix: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69628
* Ignore src/component/NetworkBlockChecker.tsx -- looks like dead code, so we could also just delete it
* findSelectedRow had an unneeded parameter. I checked that the filtered words page edit flow still seems to work.
* emoji.tsx: missing some `: string` annotations. Also TIL `.charCodeAt()` with no parameter acts the same as `.charCodeAt(0)` but TS only accepts the latter.
* Setup `make typecheck_ts` and add it to GH workflow.

I tested that `pnpm typecheck` now passes cleanly and fails if type errors are introduced inside of src/ or to vite.config.ts

leighmacdonald commented:
The `NetworkBlockChecker.tsx` will come back soon, it was just removed from use after some refactoring to the page it was on. It still needs to be re-added, so should stay.

I am however having troubles with the route naming changes. I tried testing them but I could not navigate to any forum threads nor login with those changes. Does that behave the same way on your side?

Gcommer commented:
K, NetworkBlockChecker.tsx is left it, but it's excluded in tsconfig.json due to missing packages.

My route changes were just wrong, based on https://tanstack.com/router/latest/docs/framework/react/guide/routing-concepts#index-routes. I reverted them and updated the useRouteContext hooks to have trailing slashes instead. Added some forums locally to test and they seem to work.

I now bumped the @tanstack/react-router as there were actually a few index route specific changes recently.

leighmacdonald commented:
Looks good, all working now :+1:
